In [1]:
import sys
from pathlib import Path

PROJECT_ROOT = Path().resolve().parent
sys.path.insert(0, str(PROJECT_ROOT))

from src.data.load_data import load_bank_data
from src.data.preprocess import split_data, BankFeatureEngineer, build_enhanced_pipeline
from src.models.xgboost_model import (
    train_xgboost,
    find_best_threshold,
    evaluate_xgb,
    save_xgboost,
)

In [2]:
# Train XGBoost block

# Load dataset
df = load_bank_data("../data/raw/bank-full.csv")

# Split
X_train, X_test, y_train, y_test = split_data(df)

# Detect engineered columns
fe = BankFeatureEngineer()
X_tmp = fe.fit_transform(X_train)

num_cols = X_tmp.select_dtypes(include=["int64", "float64"]).columns.tolist()
cat_cols = X_tmp.select_dtypes(include=["object", "category"]).columns.tolist()

# Calculate scale_pos_weight for imbalanced classification
pos = sum(y_train == 1)
neg = sum(y_train == 0)

scale_pos_weight = neg / pos
print("scale_pos_weight:", scale_pos_weight)


# Build preprocessing pipeline
preprocess_pipeline = build_enhanced_pipeline(num_cols, cat_cols)

# Train XGB
best_xgb, best_params, best_cv_f1 = train_xgboost(
    preprocess_pipeline,
    scale_pos_weight,
    X_train,
    y_train
)

print("Best params:", best_params)
print("Best CV F1:", best_cv_f1)

scale_pos_weight: 7.548333727251241
Fitting 5 folds for each of 64 candidates, totalling 320 fits
Best params: {'classifier__colsample_bytree': 0.9, 'classifier__learning_rate': 0.03, 'classifier__max_depth': 8, 'classifier__min_child_weight': 3, 'classifier__n_estimators': 400, 'classifier__subsample': 0.9}
Best CV F1: 0.4606087342859012


In [ ]:
#Evaluate and save model block

# Find optimal threshold
best_threshold, best_f1 = find_best_threshold(
    best_xgb,
    X_test,
    y_test
)

print("Best threshold:", best_threshold)
print("Best F1:", best_f1)

# Final evaluation
metrics = evaluate_xgb(
    best_xgb,
    X_test,
    y_test,
    best_threshold
)

print("\n--- FINAL XGBOOST ---")
for k, v in metrics.items():
    print(f"{k}: {v}")

# Save model
save_xgboost(best_xgb, best_threshold)

Best threshold: 0.609090909090909
Best F1: 0.4930102516309413

--- FINAL XGBOOST ---
Accuracy: 0.8796859449297799
AUC: 0.7974573071200372
Precision: 0.48621323529411764
Recall: 0.5
F1: 0.4930102516309413
MCC: 0.4248228671346422


### XGBoost Observations

- The dataset is highly imbalanced (~88% 'no', ~12% 'yes'), so F1-score is used as the primary evaluation metric.

- XGBoost is a gradient boosting algorithm that builds trees sequentially, improving errors from previous trees.

- Class imbalance was handled using scale_pos_weight, calculated as the ratio of negative to positive samples.

- Hyperparameter tuning selected:
  → learning_rate = 0.03 (slower learning for better generalization)
  → max_depth = 8 (balanced model complexity)
  → subsampling and column sampling to reduce overfitting

- AUC score (~0.80) indicates strong ranking ability of the model.

- Threshold tuning identified an optimal threshold (~0.60), improving F1-score.

- Final F1 score (~0.49) is the highest among all models, showing best classification performance.

- Precision (~0.49) and Recall (~0.50) are well balanced:
  → Good detection of positive class
  → Controlled false positives

- MCC (~0.42) confirms strong overall predictive performance.

Conclusion:
XGBoost provides the best performance among all models and is selected as the final model for deployment.